src/Tools/ExternalAccess/Copilot/Internal/CodeMapper/CopilotCSharpMapCodeService.cs

```diff
@@ -17,17 +19,52 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.Copilot.Internal.CodeMapper;
 [ExportLanguageService(typeof(IMapCodeService), language: LanguageNames.CSharp), Shared]
 internal sealed class CSharpMapCodeService : IMapCodeService
 {
-    private readonly ICSharpCopilotMapCodeService _service;
+    private const string CodeMapperDllName = "Microsoft.VisualStudio.Conversations.CodeMappers.CSharp, Version=0.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a";
+    private const string MapCodeServiceTypeFullName = "Microsoft.VisualStudio.Conversations.CodeMappers.CSharp.CSharpMapCodeService";
+    private const string MapCodeAsyncMethodName = "MapCodeAsync";
+
+    private readonly ICSharpCopilotMapCodeService? _service;
 
     [ImportingConstructor]
     [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-    public CSharpMapCodeService(ICSharpCopilotMapCodeService service)
+    public CSharpMapCodeService([Import(AllowDefault = true)] ICSharpCopilotMapCodeService? service)
     {
         _service = service;
     }
 
     public Task<ImmutableArray<TextChange>?> MapCodeAsync(Document document, ImmutableArray<string> contents, ImmutableArray<(Document, TextSpan)> focusLocations, CancellationToken cancellationToken)
     {
-        return _service.MapCodeAsync(document, contents, focusLocations, cancellationToken);
+        var options = new Dictionary<string, object>();
+        if (_service is not null)
+        {
+            return _service.MapCodeAsync(document, contents, focusLocations, options, cancellationToken);
+        }
+
+        return TryLoadAndInvokeViaReflectionAsync();
+
+        // The implementation of ICSharpCopilotMapCodeService is in Copilot Chat repo, which can't reference the EA package
+        // since it's shipped as a separate vsix and needs to maintain compatibility for older VS versions. So we try to call
+        // the service via reflection here until they can move to newer Roslyn version.
+        // https://github.com/dotnet/roslyn/issues/69967
+        Task<ImmutableArray<TextChange>?> TryLoadAndInvokeViaReflectionAsync()
+        {
+            try
+            {
+                var assembly = Assembly.Load(CodeMapperDllName);
+                var type = assembly.GetType(MapCodeServiceTypeFullName);
```

Comment:
I'm not as familiar with caching from reflection, but should these be cached?

Reply:
I'm not sure. I will check and fix in a follow up PR if needed. Thanks for the review!